tokio/src/sync/broadcast.rs

```diff
@@ -877,7 +877,7 @@ impl<'a, T> WaitersList<'a, T> {
 }
 
 impl<T> Shared<T> {
-    fn notify_rx<'a, 'b: 'a>(&'b self, mut tail: MutexGuard<'a, Tail>) {
+    fn notify_rx<'a, 'b: 'a>(&'b self, mut tail: RwLockWriteGuard<'a, Tail>) {
```

Comment:
This guard is only needed to synchronize with waiters that remove themselves, so technically, this only needs to be a read lock.

Reply:
At the very least, we need to convert the waiters list into the guarded version, which requires a write lock. Are you saying that we could then reacquire the lock as a read lock?

Reply:
Ah true. The write lock is required to create the guarded list, but not to traverse it.

Reply:
You can add a `downgrade` method to our wrappers around the write guards. With the `parking_lot` feature this can call into the downgrade method that parking lot provides. For std, we can fall back to drop-and-reacquire.

Reply:
It looks like we will have a data race on `waiter.queued`. `Recv::drop` now first checks `waiter.queued` with a read lock, and only upgrades it to a write lock if it needs to remove the waiter (this optimization is a big performance boost). So if we want to modify `waiter.queued` in `Shared::notify_rx`, we need a write lock.

Reply:
Oh, and we also need write access to `waiter.waker` to move it into `WakeList`, which can data race with `Receiver::recv_ref`...
In general, having only a read lock in `Shared::notify_rx` breaks the invariant that, when we access a `waiter`, we either have a write lock on tail or we have a read lock and an exclusive reference to `Recv`. Justifying safety becomes very tricky.

Reply:
Ah, yes, I assumed that `Recv::drop` only took a write lock. You will have to avoid the read lock somehow. As you mention, making `queued` atomic could work.

Reply:
Looks like making both `queued` and `waker` atomic should work. Then the following invariants can be held:

1. Accessing atomics only requires a guarantee that the containing `Recv` won't be dropped in the meantime,
2. Pushing into the list requires a read lock on tail,
3. If the list is locally owned, then removing a waiter only requires a read lock on tail,
4. If list is not locally owned, then removing a waiter requires a write lock on tail.

I will think a bit more and give it a try.

Reply:
The downside of having an `AtomicWaker` is that we will lose the [will_wake optimization](https://docs.rs/tokio/latest/src/tokio/sync/broadcast.rs.html#1098).

Reply:
You can't make `waker` atomic. However, I think this would work:

* When `queued == false`, the `Recv` future has exclusive access to `waker`.
* When `queued == true`, you must ensure that you're the only one accessing the waker to access it.

And then exclusive access to the waker is controlled by the rwlock. (With the logic that N-1 writers and 1 reader provides the same locking guarantees as N writers.)